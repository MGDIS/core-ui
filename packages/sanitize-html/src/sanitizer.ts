import sanitizeHtml from 'sanitize-html';
import { SanitizerOptions } from './types';

const DEFAULT_ALLOWED_TAGS = [...sanitizeHtml.defaults.allowedTags, 'img', 'mg-icon'];
const DEFAULT_ALLOWED_ATTRIBUTES = {
  // These is the default config of the lib
  // (copied to enforce the string type, because the lib also accepts objects for attributes)
  'a': ['href', 'target'],
  'img': ['src', 'srcset', 'alt', 'width', 'height', 'loading'],
  // Below are custom attributes
  '*': ['style', 'class', 'aria-*'],
  'mg-icon': ['icon', 'size', 'variant', 'variant-style', 'color'],
};

export class Sanitizer {
  #allowedTags: string[] = DEFAULT_ALLOWED_TAGS;
  #allowedAttributes: Record<string, string[]> = { ...DEFAULT_ALLOWED_ATTRIBUTES };

  constructor(options?: SanitizerOptions) {
    if (options?.disallowTags) {
      this.#allowedTags = this.#allowedTags.filter(tag => options.disallowTags?.indexOf(tag) === -1);
    }

    if (options?.disallowAttributes) {
      for (const elementName in options.disallowAttributes) {
        const defaultAllowedAttributes = this.#allowedAttributes[elementName];

        if (Array.isArray(defaultAllowedAttributes)) {
          this.#allowedAttributes[elementName] = defaultAllowedAttributes.filter(allowedAttribute => options.disallowAttributes?.[elementName]?.indexOf(allowedAttribute) === -1);
        }
      }
    }
  }

  /**
   * Sanitize an unsafe HTML
   * @param html - unsafe HTML
   * @returns sanitized HTML
   */
  sanitize(html: string) {
    return sanitizeHtml(html, {
      allowedTags: this.#allowedTags,
      allowedAttributes: this.#allowedAttributes,
    });
  }
}
