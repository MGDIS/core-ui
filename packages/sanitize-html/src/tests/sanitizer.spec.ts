import { describe, it, expect } from 'vitest';
import { Sanitizer } from '../sanitizer';

describe('Sanitizer', () => {
  it.each([
    ['allow aria-* everywhere', /* HTML */ `<div aria-label="test"><span aria-hidden="true"></span></div>`],
    [
      'allow image but remove unsafe properties',
      /* HTML */ `<img src="./img.png" alt="test" width="100" height="100" onload="console.log('this is not safe')" onerror="console.log('this is not safe')" />`,
    ],
    [
      'remove script tags',
      /* HTML */ `<script>
        console.log('test');
      </script>`,
    ],
    ['allow class attribute', /* HTML */ `<div class="test"></div>`],
    ['allow style attribute', /* HTML */ `<span style="color:#ff0099;background-color:#000066;text-align:right;">test</span>`],
    ['allow formatting tags', /* HTML */ `<strong>test</strong><em>test</em><sub>test</sub><sup>test</sup>`],
    [
      'allow unordered lists',
      /* HTML */ `<ul style="text-align:left;">
        <li>test</li>
        <li>test</li>
      </ul> `,
    ],
    [
      'allow ordered lists',
      /* HTML */ `<ol>
        <li>test</li>
        <li>test</li>
      </ol>`,
    ],
    ['allow links', /* HTML */ `<a href="http://localhost" title="test" target="_blank">test</a>`],
    [
      'allow tables',
      /* HTML */ `
        <table>
          <tbody>
            <tr style="height:50%;">
              <td style="width:50%;">1</td>
              <td style="width:50%;">2</td>
            </tr>
            <tr style="height:50%;">
              <td style="width:50%;">3</td>
              <td style="width:50%;">4</td>
            </tr>
          </tbody>
        </table>
      `,
    ],
    [
      'allow titles',
      /* HTML */ `<h1>Test</h1>
        <h2>test</h2>
        <h3>test</h3>`,
    ],
    [
      'allow blockquotes',
      /* HTML */ `
        <blockquote>
          <p>test<br /></p>
        </blockquote>
      `,
    ],
    ['allow mg-icon', /* HTML */ `<mg-icon icon="address-card"></mg-icon>`],
  ])('Should sanitize HTML with the default config: %s', (_description, unsafeHtml) => {
    const sanitizer = new Sanitizer();

    const sanitizedHtml = sanitizer.sanitize(unsafeHtml);

    expect(sanitizedHtml).toMatchSnapshot();
  });

  it('Should sanitize HTML and remove configured disallowed tags', () => {
    const sanitizer = new Sanitizer({ disallowTags: ['img'] });

    const sanitizedHtml = sanitizer.sanitize(/* HTML */ `
      <h1 style="color: red">Sanitize Test</h1>
      <img src="./img.png" />
      <a href="#anchor" target="_blank">a link</a>
      <script>
        console.log('this is not safe');
      </script>
    `);

    expect(sanitizedHtml).toMatchSnapshot('sanitized HTML without images');
  });

  it('Should sanitize HTML and remove configured disallowed attributes', () => {
    const sanitizer = new Sanitizer({
      disallowAttributes: {
        '*': ['style'],
        'a': ['target'],
      },
    });

    const sanitizedHtml = sanitizer.sanitize(/* HTML */ `
      <h1 style="color: red">Sanitize Test</h1>
      <img src="./img.png" />
      <a href="#anchor" target="_blank">a link</a>
      <script>
        console.log('this is not safe');
      </script>
    `);

    expect(sanitizedHtml).toMatchSnapshot('sanitized HTML without style and anchor target attributes');
  });

  it('Should ignore disallow of attributes on tags not allowed in the default config (and remove the tag)', () => {
    const sanitizer = new Sanitizer({ disallowAttributes: { 'not-an-allowed-tag': ['*'] } });

    const sanitizedHtml = sanitizer.sanitize(/* HTML */ `<not-an-allowed-tag class="test"></not-an-allowed-tag>`);

    expect(sanitizedHtml).toBe('');
  });
});
