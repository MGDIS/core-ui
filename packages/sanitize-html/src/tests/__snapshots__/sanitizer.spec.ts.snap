// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sanitizer > Should sanitize HTML and remove configured disallowed attributes > sanitized HTML without style and anchor target attributes 1`] = `
"
      <h1>Sanitize Test</h1>
      <img src="./img.png" />
      <a href="#anchor">a link</a>
      
    "
`;

exports[`Sanitizer > Should sanitize HTML and remove configured disallowed tags > sanitized HTML without images 1`] = `
"
      <h1 style="color:red">Sanitize Test</h1>
      
      <a href="#anchor" target="_blank">a link</a>
      
    "
`;

exports[`Sanitizer > Should sanitize HTML with the default config: allow aria-* everywhere 1`] = `"<div aria-label="test"><span aria-hidden="true"></span></div>"`;

exports[`Sanitizer > Should sanitize HTML with the default config: allow blockquotes 1`] = `
"
        <blockquote>
          <p>test<br /></p>
        </blockquote>
      "
`;

exports[`Sanitizer > Should sanitize HTML with the default config: allow class attribute 1`] = `"<div class="test"></div>"`;

exports[`Sanitizer > Should sanitize HTML with the default config: allow formatting tags 1`] = `"<strong>test</strong><em>test</em><sub>test</sub><sup>test</sup>"`;

exports[`Sanitizer > Should sanitize HTML with the default config: allow image but remove unsafe properties 1`] = `"<img src="./img.png" alt="test" width="100" height="100" />"`;

exports[`Sanitizer > Should sanitize HTML with the default config: allow links 1`] = `"<a href="http://localhost" target="_blank">test</a>"`;

exports[`Sanitizer > Should sanitize HTML with the default config: allow mg-icon 1`] = `"<mg-icon icon="address-card"></mg-icon>"`;

exports[`Sanitizer > Should sanitize HTML with the default config: allow ordered lists 1`] = `
"<ol>
        <li>test</li>
        <li>test</li>
      </ol>"
`;

exports[`Sanitizer > Should sanitize HTML with the default config: allow style attribute 1`] = `"<span style="color:#ff0099;background-color:#000066;text-align:right">test</span>"`;

exports[`Sanitizer > Should sanitize HTML with the default config: allow tables 1`] = `
"
        <table>
          <tbody>
            <tr style="height:50%">
              <td style="width:50%">1</td>
              <td style="width:50%">2</td>
            </tr>
            <tr style="height:50%">
              <td style="width:50%">3</td>
              <td style="width:50%">4</td>
            </tr>
          </tbody>
        </table>
      "
`;

exports[`Sanitizer > Should sanitize HTML with the default config: allow titles 1`] = `
"<h1>Test</h1>
        <h2>test</h2>
        <h3>test</h3>"
`;

exports[`Sanitizer > Should sanitize HTML with the default config: allow unordered lists 1`] = `
"<ul style="text-align:left">
        <li>test</li>
        <li>test</li>
      </ul> "
`;

exports[`Sanitizer > Should sanitize HTML with the default config: remove script tags 1`] = `""`;
